File: split-strings-by-separator-2788/Solution.java

/**
 * LeetCode problem # 2788, 'Split Strings by Separator'
 * https://leetcode.com/problems/split-strings-by-separator
 */

import java.util.*;

class Solution {
    public List<String> splitWordsBySeparator(List<String> words, char separator) {
        List<String> res = new ArrayList<>();
        for (String word : words) {
            String sep = "\\" + separator;
            String[] curr = word.split(sep);
            for (String c : curr) {
                if (c.length() > 0) {
                    res.add(c);
                } else {
                    continue;
                }
            }
        }
        return res;
    }
}

File: split-strings-by-separator-2788/solution.js
/**
 * LeetCode problem # 2788, 'Split Strings by Separator'
 * https://leetcode.com/problems/split-strings-by-separator
 */

/**
 * @param {string[]} words
 * @param {character} separator
 * @return {string[]}
 */
function splitWordsBySeparator(words, sep) {
  const res = [];
  for (const word of words) {
    for (elem of splitIntoWordsGenerator(word, sep)) {
      res.push(elem);
    }
  }
  return res;
}

function* splitIntoWordsGenerator(str, sep) {
  let curr = "";
  for (let i = 0; i < str.length; ++i) {
    if (str[i] !== sep) {
      curr += str[i];
    } else {
      if (curr.length) {
        yield curr;
        curr = "";
      }
    }
  }
  if (curr.length) {
    yield curr;
  }
}
